conn.go

```diff
@@ -1500,6 +1500,50 @@ func QuoteIdentifier(name string) string {
 	return `"` + strings.Replace(name, `"`, `""`, -1) + `"`
 }
 
+// QuoteLiteral quotes a 'literal' (e.g. a parameter, often used to pass literal
+// to DDL and other statements that do not accept parameters) to be used as part
+// of an SQL statement.  For example:
+//
+//    exp_date := pq.QuoteLiteral("2023-01-05 15:00:00Z")
+//    err := db.Exec(fmt.Sprintf("CREATE ROLE my_user VALID UNTIL %s", exp_date))
+//
+// Any single quotes in name will be escaped. Any string that contains a backslash
+// (i.e. "\")
+func QuoteLiteral(literal string) string {
+	// This follows the PostgreSQL internal algorithm for handling quoted literals
+	// which is found here:
+	// https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/backend/utils/adt/quote.c;hb=HEAD#l46
```

Comment:
We should be looking at the `libpq` implementations for inspiration, I think?

[PQescapeLiteral](https://github.com/postgres/postgres/blob/1f39a1c0641531e0462a4822f2dba904c5d4d699/src/interfaces/libpq/fe-exec.c#L3564-L3568):
```c
char *
PQescapeLiteral(PGconn *conn, const char *str, size_t len)
{
	return PQescapeInternal(conn, str, len, false);
}
```

[PQescapeInternal](https://github.com/postgres/postgres/blob/1f39a1c0641531e0462a4822f2dba904c5d4d699/src/interfaces/libpq/fe-exec.c#L3437-L3444):
```c
/*
 * Escape arbitrary strings.  If as_ident is true, we escape the result
 * as an identifier; if false, as a literal.  The result is returned in
 * a newly allocated buffer.  If we fail due to an encoding violation or out
 * of memory condition, we return NULL, storing an error message into conn.
 */
static char *
PQescapeInternal(PGconn *conn, const char *str, size_t len, bool as_ident
```

There, I see what looks like the bottom half of the function implemented in this PR:

1. Repeat any single quotes.
2. Repeat any backslashes.
3. Indicate an escape string when there are any backslashes.

What's missing here is the whitespace before the escape string's `E`. (I understand the English of the comment that justifies it, but ... 🤷‍♂️) We should do that too because, as I recall, the goal of this module is least-surprise compared to `libpq`.

Introduced in 9.0. The related thread: (maybe) https://www.postgresql.org/message-id/flat/162867790912290857m340ab1f1s76ef48c6e7241398%40mail.gmail.com

Reply:
Yeah, I agree with the above :-) I'll take a look at that + the libpq code, revise, and revise the tests as well.